File: haskell/strain/src/Strain.hs
module Strain (keep, discard) where

keep :: (a -> Bool) -> [a] -> [a]
keep _ [] = []
keep p xs = [x | x <- xs, p x]
keep _ xs = xs

discard :: (a -> Bool) -> [a] -> [a]
discard p = keep (not.p)

File: ruby/strain/strain.rb
class Array
  def keep
    return enum_for(:accumulate) unless block_given?
    each_with_object([]) { |obj, acc| acc << obj if yield(obj) }
  end

  def discard
    return enum_for(:accumulate) unless block_given?
    each_with_object([]) { |obj, acc| acc << obj unless yield(obj) }
  end
end
